### PR TITLE
BIT-5: Initial project setup

### PR DESCRIPTION
**Pull request by ElyasSalar**
The setup includes configuration and setting up the following
- eslint
- prettier
- dotenv
- nodemon
- Typescript
- winston
- mongodb
- ApiError
- ApiResponse

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+  "printWidth": 80,
+  "tabWidth": 2,
+  "semi": false,
```
Comment (HediMuhamad) [inline .prettierrc:4]: Having semicolon in the end of each statements will help our eyes to catch them better.

File: src/dao/studentsDAO.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { MongoClient } from "mongodb"
+
+let client: MongoClient
+
+export default class StudentDAO {
+  static async injectDB(connection: MongoClient) {
+    if (client instanceof MongoClient) return
```
Comment (HediMuhamad) [inline src/dao/studentsDAO.ts:7]: I'm not sure about the purpose, but as I see _client_ always is a typeof _undefined_ due to the un-initialization.

File: .prettierrc (hunk 1 above)
Comment (ElyasSalar) [reply .prettierrc:4]: I disagree. semicolons are symbols, symbols decrease the readability of code. If you look at Python, it's known for how easy it is to learn and one of the main factors is that it's near to the human language by using indentations with its few symbols and keywords. So symbols are not part of the human language. I find it more readable without it.

Comment (HediMuhamad) [reply .prettierrc:4]: Using semicolon in codes make it harder to write, easier to read, it tells us that it's the end of the statement.
Note the "dot" in the line above, what you guess when you find nothing there? We as a human use a lot of symbols in writing, not talking, and we are writing code here. 😀
About writing it down, we can use prettier to automatically adding it, let semicolons rest our eyes.

At the end, stylize as you prefer, I'm feeling well with both ways.

File: src/server.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import "dotenv/config"
+import express, { NextFunction } from "express"
+import students from "./routes/students"
```
Comment (HediMuhamad) [inline src/server.ts:3]: What about **studentRouter** for better name convention?

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+const port: number = Number(process.env.PORT) || 8000
+
+MongoClient.connect(DB_URI, MONGO_OPTIONS)
+  .then(async (client: MongoClient) => {
+    await StudentDAO.injectDB(client)
+
+    server.listen(port)
+  })
+  .catch((err: MongoNotConnectedError) => {
+    throw new Error(err.message)
```
Comment (HediMuhamad) [inline src/index.ts:15]: I think that we don't need to add these type of error to logs, logs is used to process logging not to the failure of server starting up which we exactly know what caused to the failure.

File: src/lib/logger.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { TransformableInfo } from "../types/logger.d"
+
+export const templateFunction = (info: TransformableInfo) => {
+  let template = ""
+  template += `${new Date()} `
```
Comment (HediMuhamad) [inline src/lib/logger.ts:5]: This format of date doesn't give us **MilliSecond** time, which we need when we get more than one request in a second.

**Pull request merged**